fix: preserve identity 'I' characters in PauliEvolutionGate labels

Aqil-Ahmad opened the pull request:
<!--
⚠️  If you do not respect this template, your pull request will be closed.
⚠️  Your pull request title should be short detailed and understandable for all.
⚠️  Also, please add a release note file using reno if the change needs to be documented in the release notes.
⚠️  If your pull request fixes an open issue, please link to the issue. Use "Fixes #XXXX" if this PR *fully* closes the issue XXXX.  
☢️  If you used an AI tool to code this PR, add "AI tool used: <Name and version of the tool>". For example, "AI tool used: Microsoft Copilot Chat with GPT-5". Failing to disclose the use of AI tools may result in the PR being closed without further review.  


- [ ] I have added the tests to cover my changes.
- [x] I have updated the documentation accordingly.
- [x] I have read the CONTRIBUTING document.
-->

### Summary

Fixes #15167 

This patch ensures that PauliEvolutionGate visualization labels preserve 'I' characters in the Pauli strings. Previously, these were ignored, and different gates (e.g., XXII, IXXI, IIXX) appeared identical in text circuit drawings.


### Details and comments

Tested locally with the reproduction script and the output now clearly differentiates between the gates.

Aqil-Ahmad committed:
fix: preserve identity 'I' characters in PauliEvolutionGate labels

Cryoris commented on qiskit/circuit/library/pauli_evolution.py:
```diff
@@ -335,9 +335,12 @@ def _to_sparse_op(
-            return operator[0].bit_labels()[::-1]
-        return "(" + " + ".join(term.bit_labels()[::-1] for term in operator) + ")"
+        # convert to SparsePauliOp to get full labels (with 'I' positions)
+        sp = SparsePauliOp.from_sparse_observable(operator)
```
This conversion is not for free, we shouldn't be doing that just for the label. Thinking about this, a better solution than padding with `"I"` would be to add the index of the Pauli, reflecting that the gate stores a `SparseObservable` internally.

This would mean we could show `X0 X11 + Z2` instead of `XIIIIIIIIIIX + IIIIIIIIIZII` on a 12-qubit observable. You can get this information e.g. from iterating over the observable terms
```
# if it's a SparseObservable
for term in operator:
   term.indices # array of indices
   term.bit_terms # array of the bit terms
```

Aqil-Ahmad committed:
use bit_label

Aqil-Ahmad commented:
Hi @Cryoris! can you please review

also some checks are failing. i think the failing test is expecting the old compact format, should i update it?

alexanderivrii commented:
Hi @Aqil-Ahmad, I am really sorry for the confusion, even though I have assigned to you to the issue, I did not realize that this PR and #15261 are two different PRs.

Would you be willing to finish this PR, taking @Vivek1106-04 as a co-author, and collaborating on including tests and release note?

I will defer to @Cryoris for which output format is best: the one with padded "I"s or the one with explicit indices

Aqil-Ahmad commented:
> Hi @Aqil-Ahmad, I am really sorry for the confusion, even though I have assigned to you to the issue, I did not realize that this PR and #15261 are two different PRs.
> 
> Would you be willing to finish this PR, taking @Vivek1106-04 as a co-author, and collaborating on including tests and release note?

hi @alexanderivrii, no worries at all!
i’d be happy to finish this PR and collaborate with @Vivek1106-04 and i'll include @Vivek1106 as a co-author.

Vivek1106-04 committed:
Added tests for verifying the label correctness of PauliEvolutionGate, including single term, multiple terms, and circuit display labels.

Added tests for verifying the label correctness of PauliEvolutionGate, including single term, multiple terms, and circuit display labels.

Vivek1106-04 commented:
@Aqil-Ahmad i have added tests .

Vivek1106-04 committed:
Merge branch 'main' into fix/pauli-evolution-I-labels

Aqil-Ahmad commented on test/python/circuit/library/test_evolution_gate.py:
```diff
@@ -926,6 +926,37 @@ def observable_supporting_evolution(circuit, pauli, time):
+
+    def test_single_term_label(self):
+        evo = PauliEvolutionGate(SparseObservable.from_list([("XXII", 1)]), time=1)
+        assert evo.label == "XXII"
```
you should expect X2 X3 based on the code. and so for the other cases

Aqil-Ahmad commented:
> @Aqil-Ahmad i have added tests .

thanks @Vivek1106-04! 
note that the tests will fail. based on the @Cryoris comment the tests should expect indexed notation instead of padded strings.

i'll fix these.

Aqil-Ahmad committed:
fix tests

Vivek1106-04 committed:
Merge branch 'main' into fix/pauli-evolution-I-labels

Vivek1106-04 commented:
@Aqil-Ahmad thanks for updating my tests . it's nice working with @Aqil-Ahmad , @alexanderivrii , @Cryoris

alexanderivrii commented on test/python/circuit/library/test_evolution_gate.py:
```diff
@@ -926,6 +926,37 @@ def observable_supporting_evolution(circuit, pauli, time):
+            ],
+            time=1,
+        )
+        expected_label = "exp(-it (['(X0 X1 + Y1 Y2 + Z2 Z3)', 'X2 X3']))"
```
Question for @Cryoris: do you prefer this output or the one one without the extra `'` symbols: `expected_label = "exp(-it ([(X0 X1 + Y1 Y2 + Z2 Z3), X2 X3]))"`?

alexanderivrii commented:
Thank you very much! This is very-very close. Modulo yet an additional tweak to the output that we may want to do (I have left a question for @Cryoris), there is a minor formatting problem (so, just run `black`), and a test failure for an existing testcase `test_labels_and_name` (since we now use the indexed notation, the test needs to be updated). Also, please add a release note.

Aqil-Ahmad commented:
> Thank you very much! This is very-very close. Modulo yet an additional tweak to the output that we may want to do (I have left a question for @Cryoris), there is a minor formatting problem (so, just run `black`), and a test failure for an existing testcase `test_labels_and_name` (since we now use the indexed notation, the test needs to be updated). Also, please add a release note.

thanks @alexanderivrii! i'll push the updates shortly.

Aqil-Ahmad committed:
add release note and update test

Aqil-Ahmad committed:
Merge branch 'fix/pauli-evolution-I-labels' of https://github.com/Aqil-Ahmad/qiskit into fix/pauli-evolution-I-labels

Aqil-Ahmad commented:
Hi @Vivek1106-04! i've updated the test and added a release note. check the changes and feel free to make any fixes if needed.

alexanderivrii commented on releasenotes/notes/improve-pauli-evolution-labels-76cb9cb159186b26.yaml:
```diff
@@ -0,0 +1,7 @@
+fixes:
+  - |
+    Fixed label generation for :class:`~.PauliEvolutionGate` with 
+    :class:`~.SparseObservable` operators. Labels now display Pauli operators 
```
A super minor-nitpick: technically we do not need "~" here, as a tilde prefix shortens the link text to the last component of the target, see https://www.sphinx-doc.org/en/master/usage/referencing.html, and we already have the link text consisting of a single component.

```suggestion
    Fixed label generation for :class:`.PauliEvolutionGate` with 
    :class:`.SparseObservable` operators. Labels now display Pauli operators 
```

alexanderivrii commented on releasenotes/notes/improve-pauli-evolution-labels-76cb9cb159186b26.yaml:
```diff
@@ -0,0 +1,7 @@
+    Fixed label generation for :class:`~.PauliEvolutionGate` with 
+    :class:`~.SparseObservable` operators. Labels now display Pauli operators 
+    with qubit indices (e.g., "X0 X1") instead of the concatenated string format 
+    (e.g., "XXII"), improving clarity when distinguishing between different operators.
```
Maybe change the indices as follows, to more clearly show the problem fixed by this PR?

```suggestion
    with qubit indices (e.g., "X0 X2") instead of the concatenated string format 
    (e.g., "XX"), improving clarity when distinguishing between different operators.
```

alexanderivrii reviewed (commented):
Just two minor suggestions about the release notes, after this, it's good to merge.

Aqil-Ahmad committed:
minor changes

Vivek1106-04 commented:
@Aqil-Ahmad @alexanderivrii it's nice to work with you all , this was my first opensource work .
